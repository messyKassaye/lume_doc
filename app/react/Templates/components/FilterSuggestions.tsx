import { connect } from 'react-redux';
import React, { Component } from 'react';

import { Translate } from 'app/I18N';
import { TemplateSchema } from 'shared/types/templateType';
import { PropertySchema } from 'shared/types/commonTypes';
import { IImmutable } from 'shared/types/Immutable';
import { getCompatibleTypes } from 'shared/propertyTypes';
import { SimilarProperty, TemplateProperty } from './SimilarProperty';

interface MatchedProperty {
  template: string;
  property: PropertySchema;
}

class SharedProperties extends Component<FilterSuggestionsProps> {
  getRelationTypeName(relationTypeId: string | undefined) {
    const relationType = relationTypeId
      ? this.props.relationTypes.toJS().find((r: any) => r._id === relationTypeId)
      : undefined;
    return relationType ? relationType.name : undefined;
  }

  getThesauriName(thesauriId: string | undefined) {
    const thesaurus = thesauriId
      ? this.props.thesauris.toJS().find((thesauri: any) => thesauri._id === thesauriId)
      : undefined;
    return thesaurus ? thesaurus.name : undefined;
  }

  findSameLabelProperties(label: string, templates: TemplateSchema[]): MatchedProperty[] {
    return templates
      .filter(template => template._id !== this.props.templateId && template.properties)
      .map(template => {
        const property = (template.properties || []).find(
          prop => prop.label.trim().toLowerCase() === label.trim().toLowerCase()
        );
        return (property
          ? { template: template.name, property }
          : { template: undefined }) as MatchedProperty;
      })
      .filter(prop => prop.template !== undefined);
  }

  render() {
    const { type, content, relationType } = this.props;

    const similarProperties: TemplateProperty[] = this.findSameLabelProperties(
      this.props.label,
      this.props.templates.toJS()
    ).map((propertyMatch: MatchedProperty) => {
      const { property } = propertyMatch;
      return {
        ...propertyMatch,
        typeConflict: !getCompatibleTypes(type).includes(property.type),
        relationConflict: relationType && property.relationType !== relationType,
        contentConflict: property.content !== content,
        type: property.type,
        relationTypeName: this.getRelationTypeName(property.relationType),
        thesaurusName: this.getThesauriName(property.content),
      } as TemplateProperty;
    });

    const thisProperty: TemplateProperty = {
      template: `${this.props.templateName} (this template)`,
      type,
      relationTypeName: this.getRelationTypeName(relationType),
      thesaurusName: this.getThesauriName(content),
      typeConflict: false,
      contentConflict: false,
      relationConflict: false,
    };

    const templatesWithSameLabelProperties = [thisProperty, ...similarProperties];
    const hasContent = templatesWithSameLabelProperties.find(prop => prop.thesaurusName);
    return (
      <>
        <label className="suggestions-label">
          <Translate>
            Properties from other templates in the collection using the same label
          </Translate>
        </label>
        <table className="table">
          <thead>
            <tr>
              <th>
                <Translate>Template</Translate>
              </th>
              <th>
                <Translate>Type</Translate>
              </th>
              {hasContent && (
                <th>
                  <Translate>Thesauri</Translate>/<Translate>Entity</Translate>
                </th>
              )}
            </tr>
          </thead>
          <tbody>
            {templatesWithSameLabelProperties.map((templateProperty, index) => (
              <SimilarProperty key={index} templateProperty={templateProperty} />
            ))}
          </tbody>
        </table>
      </>
    );
  }
}

export type FilterSuggestionsProps = {
  index: number;
  label: string;
  type: PropertySchema['type'];
  filter: any;
  templateName: string;
  templateId: string | { [k: string]: any } | undefined;
  templates: IImmutable<TemplateSchema[]>;
  thesauris: IImmutable<[]>;
  relationTypes: IImmutable<[]>;
  content: string;
  relationType: string;
};

export function mapStateToProps(state: any, props: FilterSuggestionsProps) {
  const propertySchemaElement = state.template.data.properties[props.index];
  const { relationTypes } = state;
  return {
    templates: state.templates,
    thesauris: state.thesauris,
    relationTypes,
    templateName: state.template.data.name,
    templateId: state.template.data._id,
    type: propertySchemaElement.type,
    filter: propertySchemaElement.filter,
    label: propertySchemaElement.label,
    content: propertySchemaElement.content,
    relationType: propertySchemaElement.relationType,
  };
}

export const FilterSuggestions = connect(mapStateToProps)(SharedProperties);
