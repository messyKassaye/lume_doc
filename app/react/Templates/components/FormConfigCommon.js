import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Field } from 'react-redux-form';
import { connect } from 'react-redux';

import PrioritySortingLabel from './PrioritySortingLabel';

export class FormConfigCommon extends Component {
  getZeroIndex() {
    const { index, data } = this.props;
    const baseZeroIndex = index + data.commonProperties.length;
    return baseZeroIndex;
  }

  renderTitleField() {
    const { index, formState } = this.props;
    let labelClass = 'form-group';
    const labelKey = `commonProperties.${this.getZeroIndex()}.label`;
    const requiredLabel = formState.$form.errors[`${labelKey}.required`];
    const duplicatedLabel = formState.$form.errors[`${labelKey}.duplicated`];
    if (requiredLabel || duplicatedLabel) {
      labelClass += ' has-error';
    }

    return (
      <div className={labelClass}>
        <label htmlFor={`label${index}`}>Name</label>
        <Field model={`template.data.commonProperties[${this.getZeroIndex()}].label`}>
          <input id={`label${index}`} className="form-control" />
        </Field>
      </div>
    );
  }

  render() {
    const { index, data } = this.props;
    const property = data.commonProperties[this.getZeroIndex()];

    return (
      <div>
        {property.name === 'title' && this.renderTitleField()}
        <Field model={`template.data.commonProperties[${this.getZeroIndex()}].prioritySorting`}>
          <input id={`prioritySorting${index}`} type="checkbox" />
          &nbsp;
          <PrioritySortingLabel htmlFor={`prioritySorting${index}`} />
        </Field>
      </div>
    );
  }
}

FormConfigCommon.propTypes = {
  data: PropTypes.shape({ commonProperties: PropTypes.array }).isRequired,
  formState: PropTypes.shape({ $form: PropTypes.object }).isRequired,
  index: PropTypes.number,
};

export function mapStateToProps({ template }) {
  return {
    data: template.data,
    formState: template.formState,
  };
}

export default connect(mapStateToProps)(FormConfigCommon);
