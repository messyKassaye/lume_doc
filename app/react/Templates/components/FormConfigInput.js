import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Field } from 'react-redux-form';
import { connect } from 'react-redux';
import { Icon } from 'app/UI';
import { Translate } from 'app/I18N';
import { checkErrorsOnLabel } from '../utils/checkErrorsOnLabel';

import PropertyConfigOptions from './PropertyConfigOptions';

export class FormConfigInput extends Component {
  render() {
    const { index, labelHasError, type, canBeFilter } = this.props;

    return (
      <div>
        <div className={`form-group${labelHasError ? ' has-error' : ''}`}>
          <label>Name</label>
          <Field model={`template.data.properties[${index}].label`}>
            <input className="form-control" />
          </Field>
        </div>
        {type === 'geolocation' && (
          <div className="geolocation-grouping-alert">
            <Icon icon="info-circle" />
            <p>
              <Translate>Adjacent geolocation properties will render on the same map</Translate>.
            </p>
          </div>
        )}
        <PropertyConfigOptions index={index} type={type} canBeFilter={canBeFilter} />
      </div>
    );
  }
}

FormConfigInput.defaultProps = {
  canBeFilter: true,
  labelHasError: false,
};

FormConfigInput.propTypes = {
  canBeFilter: PropTypes.bool,
  index: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  labelHasError: PropTypes.bool,
};

export function mapStateToProps(state, props) {
  return {
    labelHasError: checkErrorsOnLabel(state, props),
  };
}

export default connect(mapStateToProps)(FormConfigInput);
