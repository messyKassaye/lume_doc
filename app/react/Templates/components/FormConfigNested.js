import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { setNestedProperties } from 'app/Templates/actions/templateActions';
import { Field } from 'react-redux-form';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Icon } from 'UI';
import ViolatedArticlesNestedProperties from './ViolatedArticlesNestedProperties';
import PropertyConfigOptions from './PropertyConfigOptions';
import { checkErrorsOnLabel } from '../utils/checkErrorsOnLabel';

export class FormConfigNested extends Component {
  constructor(props) {
    super(props);
    props.setNestedProperties(props.index, Object.keys(ViolatedArticlesNestedProperties));
  }

  contentValidation() {
    return { required: val => val.trim() !== '' };
  }

  render() {
    const { index, type, labelHasError } = this.props;

    return (
      <div>
        <div className={`form-group${labelHasError ? ' has-error' : ''}`}>
          <label>Label</label>
          <Field model={`template.data.properties[${index}].label`}>
            <input className="form-control" />
          </Field>
        </div>

        <Field model={`template.data.properties[${index}].required`}>
          <input id={`required${this.props.index}`} type="checkbox" />
          &nbsp;
          <label className="property-label" htmlFor={`required${this.props.index}`}>
            Required property
            <span className="property-help">
              <Icon icon="question-circle" />
              <div className="property-description">
                You won&#39;t be able to publish a document if this property is empty.
              </div>
            </span>
          </label>
        </Field>
        <PropertyConfigOptions index={index} type={type} />
      </div>
    );
  }
}

FormConfigNested.defaultProps = {
  labelHasError: false,
};

FormConfigNested.propTypes = {
  thesauris: PropTypes.object,
  index: PropTypes.number,
  formKey: PropTypes.string,
  setNestedProperties: PropTypes.func,
  type: PropTypes.string.isRequired,
  labelHasError: PropTypes.bool,
};

export function mapStateToProps(state, props) {
  return {
    labelHasError: checkErrorsOnLabel(state, props),
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ setNestedProperties }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(FormConfigNested);
