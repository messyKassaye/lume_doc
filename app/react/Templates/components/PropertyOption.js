/**
 * @jest-environment jsdom
 */
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { DragSource } from 'react-dnd';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Icon } from 'UI';

import { removeProperty, addProperty } from 'app/Templates/actions/templateActions';
import Icons from './Icons';

export class PropertyOption extends Component {
  constructor(props) {
    super(props);
    this.addProperty = this.addProperty.bind(this);
  }

  addProperty() {
    const { disabled, label, type, addProperty: addPropertyAction } = this.props;
    if (!disabled) {
      addPropertyAction({ label, type });
    }
  }

  render() {
    const { connectDragSource, label, disabled, type } = this.props;
    const iconClass = Icons[type] || 'font';
    const liClass = `list-group-item${disabled ? ' disabled' : ''}`;
    return connectDragSource(
      <li className={liClass}>
        <button type="button" onClick={this.addProperty}>
          <Icon icon="plus" />
        </button>
        <span>
          <Icon icon={iconClass} /> {label}
        </span>
      </li>
    );
  }
}

PropertyOption.propTypes = {
  connectDragSource: PropTypes.func.isRequired,
  addProperty: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
};

const optionSource = {
  beginDrag({ ...props }) {
    return props;
  },

  canDrag(props) {
    return !props.disabled;
  },

  endDrag(props, monitor) {
    const item = monitor.getItem();
    const dropResult = monitor.getDropResult();
    if (!dropResult && item.index) {
      props.removeProperty(item.index);
    }
  },
};

const dragSource = DragSource('METADATA_OPTION', optionSource, connector => ({
  connectDragSource: connector.dragSource(),
}))(PropertyOption);

export { dragSource };

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ removeProperty, addProperty }, dispatch);
}

export default connect(null, mapDispatchToProps, null, { withRef: true })(dragSource);
