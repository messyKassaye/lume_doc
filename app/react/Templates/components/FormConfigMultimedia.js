import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Field } from 'react-redux-form';
import { connect } from 'react-redux';
import { Icon } from 'UI';

import { t } from 'app/I18N';
import { Select } from 'app/ReactReduxForms';
import Tip from 'app/Layout/Tip';

import { FilterSuggestions } from 'app/Templates/components/FilterSuggestions';
import PropertyConfigOption from './PropertyConfigOption';
import { checkErrorsOnLabel } from '../utils/checkErrorsOnLabel';

const style = index => (
  <div>
    <div className="form-group">
      <label>{t('System', 'Style')}</label>
      <Select
        model={`template.data.properties[${index}].style`}
        options={[
          { _id: 'contain', name: 'Fit' },
          { _id: 'cover', name: 'Fill' },
        ]}
        optionsLabel="name"
        optionsValue="_id"
      />
    </div>
    <div className="protip">
      <p>
        <b>Fit</b> will show the entire media inside the container.
        <br />
        <b>Fill</b> will attempt to fill the container, using it&#39;s entire width. In cards,
        cropping is likely to occur.
      </p>
    </div>
  </div>
);

class FormConfigMultimedia extends Component {
  render() {
    const {
      index,
      canShowInCard,
      helpText,
      canSetStyle,
      canBeRequired,
      labelHasError,
    } = this.props;

    return (
      <div>
        <div className={`form-group${labelHasError ? ' has-error' : ''}`}>
          <label>Name</label>
          <Field model={`template.data.properties[${index}].label`}>
            <input className="form-control" />
          </Field>
        </div>

        {helpText && (
          <div className="protip">
            <Icon icon="lightbulb" />
            <span>{helpText}</span>
          </div>
        )}

        <PropertyConfigOption
          label="Hide label"
          model={`template.data.properties[${index}].noLabel`}
        >
          <Tip>This property will be shown without the label.</Tip>
        </PropertyConfigOption>
        <PropertyConfigOption
          label="Full width"
          model={`template.data.properties[${index}].fullWidth`}
        >
          <Tip>This property will be shown using all the width available.</Tip>
        </PropertyConfigOption>
        {canBeRequired && (
          <PropertyConfigOption
            label="Required property"
            model={`template.data.properties[${index}].required`}
          >
            <Tip>You won&#39;t be able to save a document if this property is empty.</Tip>
          </PropertyConfigOption>
        )}
        {canShowInCard && (
          <PropertyConfigOption
            label="Show in cards"
            model={`template.data.properties[${index}].showInCard`}
          >
            <Tip>This property will appear in the library cards as part of the basic info.</Tip>
          </PropertyConfigOption>
        )}

        {canSetStyle && style(index)}
        <FilterSuggestions index={index} />
      </div>
    );
  }
}

FormConfigMultimedia.defaultProps = {
  canShowInCard: true,
  canSetStyle: true,
  canBeRequired: true,
  labelHasError: false,
  helpText: '',
};

FormConfigMultimedia.propTypes = {
  canShowInCard: PropTypes.bool,
  canSetStyle: PropTypes.bool,
  canBeRequired: PropTypes.bool,
  helpText: PropTypes.string,
  index: PropTypes.number.isRequired,
  labelHasError: PropTypes.bool,
};

export function mapStateToProps(state, props) {
  return {
    labelHasError: checkErrorsOnLabel(state, props),
  };
}

export default connect(mapStateToProps)(FormConfigMultimedia);
