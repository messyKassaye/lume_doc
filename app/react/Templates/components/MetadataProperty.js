import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { DragSource, DropTarget } from 'react-dnd';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { editProperty } from 'app/Templates/actions/uiActions';
import { showModal } from 'app/Modals/actions/modalActions';
import { reorderProperty, addProperty } from 'app/Templates/actions/templateActions';
import ShowIf from 'app/App/ShowIf';
import { Icon } from 'UI';
import { StateSelector } from 'app/Review/components/StateSelector';
import { createSelector } from 'reselect';

import FormConfigInput from './FormConfigInput';
import FormConfigSelect from './FormConfigSelect';
import FormConfigRelationship from './FormConfigRelationship';
import FormConfigRelationshipFilter from './FormConfigRelationshipFilter';
import FormConfigNested from './FormConfigNested';
import FormConfigCommon from './FormConfigCommon';
import FormConfigMultimedia from './FormConfigMultimedia';
import Icons from './Icons';

const isLabelDuplicated = (index, template, formState) => {
  const commonPropIndex = index + template.commonProperties.length;
  return (
    Boolean(formState.$form.errors[`properties.${index}.label.duplicated`]) ||
    Boolean(formState.$form.errors[`commonProperties.${commonPropIndex}.label.duplicated`])
  );
};

const isErrorOnThisField = (error, index, isCommonProperty, template) => {
  const commonPropIndex = index + template.commonProperties.length;
  const [errorRoot, errorIndex] = error.split('.');
  return errorRoot === 'commonProperties'
    ? errorIndex === commonPropIndex.toString() && isCommonProperty
    : errorIndex === index.toString() && !isCommonProperty;
};

export class MetadataProperty extends Component {
  renderForm() {
    const { type, index } = this.props;
    let defaultInput = <FormConfigInput type={type} index={index} />;

    if (this.props.isCommonProperty) {
      return <FormConfigCommon index={index} type={type} />;
    }
    if (type === 'relationship') {
      defaultInput = <FormConfigRelationship index={index} type={type} />;
    }
    if (type === 'relationshipfilter') {
      defaultInput = <FormConfigRelationshipFilter index={index} type={type} />;
    }
    if (type === 'select' || type === 'multiselect') {
      defaultInput = <FormConfigSelect index={index} type={type} />;
    }
    if (type === 'nested') {
      defaultInput = <FormConfigNested index={index} type={type} />;
    }
    if (type === 'media' || type === 'image' || type === 'preview') {
      defaultInput = (
        <FormConfigMultimedia
          type={type}
          index={index}
          canSetStyle={type === 'image' || type === 'preview'}
          canBeRequired={type !== 'preview'}
        />
      );
    }
    if (type === 'geolocation' || type === 'link') {
      defaultInput = <FormConfigInput type={type} index={index} canBeFilter={false} />;
    }
    return defaultInput;
  }

  render() {
    const {
      label,
      connectDragSource,
      isDragging,
      connectDropTarget,
      uiState,
      index,
      localID,
      inserting,
      hasErrors,
      submitFailed,
    } = this.props;
    const { editingProperty } = uiState.toJS();

    let propertyClass = 'list-group-item';
    if (isDragging || inserting) {
      propertyClass += ' dragging';
    }

    if (hasErrors && submitFailed) {
      propertyClass += ' error';
    }

    const iconClass = Icons[this.props.type] || 'font';
    const beingEdited = editingProperty === localID;

    const property = (
      <div className={propertyClass}>
        <span className="property-name">
          <Icon icon={this.props.isCommonProperty ? 'lock' : 'bars'} fixedWidth />
          <Icon icon={iconClass} fixedWidth />
          {label !== null ? (
            label
          ) : (
            <StateSelector
              propertyLabel={createSelector(
                state =>
                  state.template.data.properties[this.props.index]
                    ? state.template.data.properties[this.props.index].label
                    : '',
                value => value
              )}
            >
              {({ propertyLabel }) => propertyLabel}
            </StateSelector>
          )}
        </span>
        <div className="list-group-item-actions">
          {this.props.isLabelDuplicated && (
            <span className="validation-error">
              <Icon icon="exclamation-triangle" /> Duplicated label
            </span>
          )}
          {this.props.isRelationDuplicated && (
            <span className="validation-error">
              <Icon icon="exclamation-triangle" /> Relationship fields must have diferent
              relationship or diferent type of entity.
            </span>
          )}
          <button
            type="button"
            className="btn btn-default btn-xs property-edit"
            onClick={() => this.props.editProperty(beingEdited ? null : localID)}
          >
            <Icon icon="pencil-alt" /> Edit
          </button>
          {!this.props.isCommonProperty && (
            <button
              type="button"
              className="btn btn-danger btn-xs property-remove"
              onClick={() => this.props.removeProperty('RemovePropertyModal', index)}
            >
              <Icon icon="trash-alt" /> Delete
            </button>
          )}
        </div>
      </div>
    );

    if (this.props.isCommonProperty) {
      return (
        <li>
          {property}
          <ShowIf if={beingEdited && !isDragging}>
            <div className={`propery-form${editingProperty === localID ? ' expand' : ''}`}>
              {this.renderForm()}
            </div>
          </ShowIf>
        </li>
      );
    }

    return connectDropTarget(
      <li>
        {connectDragSource(property)}
        <ShowIf if={beingEdited && !isDragging}>
          <div className={`propery-form${editingProperty === localID ? ' expand' : ''}`}>
            {this.renderForm()}
          </div>
        </ShowIf>
      </li>
    );
  }
}

MetadataProperty.defaultProps = {
  label: null,
  submitFailed: false,
  hasErrors: false,
  isLabelDuplicated: false,
  isRelationDuplicated: false,
};

MetadataProperty.propTypes = {
  connectDragSource: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  submitFailed: PropTypes.bool,
  hasErrors: PropTypes.bool,
  isLabelDuplicated: PropTypes.bool,
  isRelationDuplicated: PropTypes.bool,
  index: PropTypes.number.isRequired,
  isDragging: PropTypes.bool.isRequired,
  localID: PropTypes.any.isRequired,
  type: PropTypes.string,
  label: PropTypes.string,
  isCommonProperty: PropTypes.bool,
  inserting: PropTypes.bool,
  removeProperty: PropTypes.func,
  uiState: PropTypes.object,
  editProperty: PropTypes.func,
};

const target = {
  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;
    const item = monitor.getItem();
    if (props.localID === item.editingProperty) {
      props.editProperty(null);
    }
    if (typeof dragIndex === 'undefined') {
      item.inserting = true;
      item.index = 0;
      props.addProperty({ label: item.label, type: item.type, inserting: true }, 0);
      return;
    }

    if (dragIndex === hoverIndex) {
      return;
    }

    if (item.alreadyReordered) {
      item.alreadyReordered = false;
      return;
    }

    props.reorderProperty(dragIndex, hoverIndex);
    item.index = hoverIndex;
    item.alreadyReordered = true;
  },
};

const dropTarget = DropTarget(['METADATA_PROPERTY', 'METADATA_OPTION'], target, connector => ({
  connectDropTarget: connector.dropTarget(),
}))(MetadataProperty);

const source = {
  beginDrag(props) {
    return {
      index: props.index,
      label: props.label,
      type: props.type,
      editingProperty: props.uiState.get('editingProperty'),
    };
  },
  endDrag(props, monitor) {
    const item = monitor.getItem();
    props.editProperty(item.editingProperty);
  },
};

const dragSource = DragSource('METADATA_PROPERTY', source, (connector, monitor) => ({
  connectDragSource: connector.dragSource(),
  isDragging: monitor.isDragging(),
}))(dropTarget);

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    { removeProperty: showModal, reorderProperty, addProperty, editProperty },
    dispatch
  );
}

const mapStateToProps = ({ template }, ownProps) => ({
  uiState: template.uiState,
  hasErrors: Object.keys(template.formState.$form.errors || {}).reduce(
    (result, error) =>
      result ||
      (isErrorOnThisField(error, ownProps.index, ownProps.isCommonProperty, template.data) &&
        template.formState.$form.errors[error]),
    false
  ),
  isLabelDuplicated: isLabelDuplicated(ownProps.index, template.data, template.formState),
  isRelationDuplicated: Boolean(
    template.formState.$form.errors[`properties.${ownProps.index}.relationType.duplicated`]
  ),
  submitFailed: template.formState.$form.submitFailed,
});

export { dragSource, dropTarget };

export default connect(mapStateToProps, mapDispatchToProps)(dragSource);
