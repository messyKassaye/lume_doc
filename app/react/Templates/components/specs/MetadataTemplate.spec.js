/**
 * @jest-environment jsdom
 */
import React, { Component } from 'react';
import TestUtils from 'react-dom/test-utils';
import TestBackend from 'react-dnd-test-backend';
import { DragDropContext } from 'react-dnd';
import { Provider } from 'react-redux';
import Immutable from 'immutable';
import { shallow } from 'enzyme';
import { modelReducer, formReducer, Field, Control } from 'react-redux-form';
import { combineReducers, createStore } from 'redux';
import api from 'app/Templates/TemplatesAPI';

import {
  MetadataTemplate,
  dropTarget,
  mapStateToProps,
} from 'app/Templates/components/MetadataTemplate';
import MetadataProperty from 'app/Templates/components/MetadataProperty';
import { dragSource } from 'app/Templates/components/PropertyOption';

function sourceTargetTestContext(Target, Source, actions) {
  return DragDropContext(TestBackend)(
    class TestContextContainer extends Component {
      render() {
        const identity = x => x;
        const properties = [
          { label: 'childTarget', localID: 'childId', inserting: true, type: 'text' },
        ];
        const commonProperties = [];
        const templates = Immutable.fromJS([]);
        const targetProps = {
          properties,
          commonProperties,
          templates,
          connectDropTarget: identity,
          formState: { fields: {}, errors: {} },
          backUrl: 'url',
          saveTemplate: jasmine.createSpy('saveTemplate'),
          defaultColor: '#112233',
        };
        const sourceProps = {
          label: 'source',
          type: 'type',
          index: 2,
          localID: 'source',
          addProperty: () => {},
          connectDragSource: identity,
          formState: { fields: {}, errors: {} },
        };

        return (
          <div>
            <Target {...targetProps} {...actions} />
            <Source {...sourceProps} />
          </div>
        );
      }
    }
  );
}

describe('MetadataTemplate', () => {
  function renderComponent(ComponentToRender, props = {}, properties = []) {
    let result;
    const formModel = {
      name: '',
      properties,
      commonProperties: [{ isCommonProperty: true, label: 'Title' }],
    };
    props.properties = properties;
    props.backUrl = 'url';
    const initialData = {
      template: { data: formModel },
      form: { template: {} },
      templates: Immutable.fromJS({ templates: [] }),
      modals: Immutable.fromJS({}),
    };
    const store = createStore(
      combineReducers({
        template: combineReducers({
          data: modelReducer('template.data', formModel),
          formState: formReducer('template.data'),
          uiState: () => Immutable.fromJS({ editProperty: '' }),
        }),
        templates: () => Immutable.fromJS([]),
        form: () => initialData.form,
        modals: () => initialData.modals,
      }),
      initialData
    );
    TestUtils.renderIntoDocument(
      <Provider store={store}>
        <ComponentToRender ref={ref => (result = ref)} {...props} index={1} />
      </Provider>
    );
    return result;
  }

  let props;
  beforeEach(() => {
    props = {
      backUrl: '',
      _id: '123',
      commonProperties: [],
      properties: [],
      connectDropTarget: x => x,
      formState: { fields: {} },
      templates: Immutable.fromJS([]),
      saveTemplate: jasmine.createSpy('saveTemplate'),
      defaultColor: '#112233',
    };
  });

  describe('render()', () => {
    it('should disable send button when saving the template', () => {
      let component = shallow(<MetadataTemplate {...props} />);
      expect(component.find('button').props().disabled).toBe(false);

      props.savingTemplate = true;
      component = shallow(<MetadataTemplate {...props} />);
      expect(component.find('button').props().disabled).toBe(true);
    });

    it('should render the template name field', () => {
      const component = shallow(<MetadataTemplate {...props} />);
      expect(component.find(Field).getElements()[0].props.model).toBe('.name');
    });

    it('should render template color field', () => {
      const component = shallow(<MetadataTemplate {...props} />);
      expect(component.find(Control).first()).toMatchSnapshot();
    });

    describe('when fields is empty', () => {
      it('should render a blank state', () => {
        const component = shallow(<MetadataTemplate {...props} />);
        expect(component.find('.no-properties').length).toBe(1);
      });
    });

    describe('when it has commonProperties', () => {
      it('should render all commonProperties as MetadataProperty', () => {
        props.commonProperties = [
          { label: 'country', type: 'text', _id: '1' },
          { label: 'author', type: 'text', _id: '2' },
        ];
        const component = shallow(<MetadataTemplate {...props} />);
        expect(component.find(MetadataProperty).length).toBe(2);
        expect(
          component
            .find(MetadataProperty)
            .at(0)
            .props().index
        ).toBe(-2);
        expect(
          component
            .find(MetadataProperty)
            .at(1)
            .props().index
        ).toBe(-1);
      });
    });

    describe('when it has properties', () => {
      it('should render all properties as MetadataProperty', () => {
        props.properties = [
          { label: 'country', type: 'text', _id: '1' },
          { label: 'author', type: 'text', _id: '2' },
        ];
        props.commonProperties = [];
        const component = shallow(<MetadataTemplate {...props} />);
        expect(component.find(MetadataProperty).length).toBe(2);
      });
    });
  });

  describe('onSubmit', () => {
    it('should trim the properties labels and then call props.saveTemplate', async () => {
      spyOn(api, 'validateMapping').and.returnValue({ errors: [], valid: true });
      const component = shallow(<MetadataTemplate {...props} />);
      const template = { properties: [{ label: ' trim me please ' }] };
      await component.instance().onSubmit(template);
      expect(props.saveTemplate).toHaveBeenCalledWith({
        properties: [{ label: 'trim me please' }],
      });
    });

    describe('when the mapping has con flicts', () => {
      it('should ask for a reindex', async () => {
        spyOn(api, 'validateMapping').and.returnValue({
          errors: [{ name: 'Date of birth' }],
          valid: false,
        });
        const context = { confirm: jasmine.createSpy('confirm') };

        const component = shallow(<MetadataTemplate {...props} />, { context });
        const template = { properties: [{ name: 'dob', type: 'date', label: 'Date of birth' }] };
        await component.instance().onSubmit(template);
        context.confirm.calls.mostRecent().args[0].accept();
        expect(props.saveTemplate).toHaveBeenCalledWith({
          properties: [{ name: 'dob', type: 'date', label: 'Date of birth' }],
          reindex: true,
        });
      });
    });
  });

  describe('dropTarget', () => {
    const actions = jasmine.createSpyObj('actions', ['inserted', 'addProperty']);
    let backend;
    let component;
    let monitor;

    beforeEach(() => {
      const TestDragAndDropContext = sourceTargetTestContext(dropTarget, dragSource, actions);
      component = renderComponent(TestDragAndDropContext, {}, [{ label: '', type: 'text' }]);
      backend = component.getManager().getBackend();
      monitor = component.getManager().getMonitor();
    });

    describe('when droping', () => {
      it('should addField on the last index', () => {
        const target = TestUtils.findRenderedComponentWithType(component, dropTarget);
        const source = TestUtils.findRenderedComponentWithType(component, dragSource);

        backend.simulateBeginDrag([source.getHandlerId()]);
        backend.simulateHover([target.getHandlerId()]);
        backend.simulateDrop();

        const lastIndex = 1;
        expect(actions.addProperty).toHaveBeenCalledWith(
          { label: 'source', type: 'type' },
          lastIndex
        );
      });
    });

    describe('when droping a property that is being inserted', () => {
      it('should updateProperty removing "inserting" flag on item index', () => {
        const target = TestUtils.findRenderedComponentWithType(component, dropTarget);
        const source = TestUtils.findRenderedComponentWithType(component, dragSource);
        const index = 0;

        backend.simulateBeginDrag([source.getHandlerId()]);
        monitor.getItem().index = index;
        backend.simulateHover([target.getHandlerId()]);
        backend.simulateDrop();

        expect(actions.inserted).toHaveBeenCalledWith(index);
      });
    });
  });

  describe('mapStateToProps', () => {
    it('should select next available template color as defaultColor for new template', () => {
      const template = { data: {}, uiState: Immutable.fromJS({}) };
      const templates = Immutable.fromJS([{ _id: 'id1' }, { _id: 'id2' }, { _id: 'id3' }]);
      const res = mapStateToProps({ template, templates }, {});
      expect(res.defaultColor).toMatchSnapshot();
    });
    it('should select defaultColor based on template index if template already exists', () => {
      const template = { data: { _id: 'id2' }, uiState: Immutable.fromJS({}) };
      const templates = Immutable.fromJS([{ _id: 'id1' }, { _id: 'id2' }, { _id: 'id3' }]);
      const res = mapStateToProps({ template, templates }, {});
      expect(res.defaultColor).toMatchSnapshot();
    });
  });
});
