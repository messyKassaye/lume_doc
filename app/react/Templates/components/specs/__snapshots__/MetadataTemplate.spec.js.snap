// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataTemplate mapStateToProps should select defaultColor based on template index if template already exists 1`] = `"#A03AB1"`;

exports[`MetadataTemplate mapStateToProps should select next available template color as defaultColor for new template 1`] = `"#A03AB1"`;

exports[`MetadataTemplate render() should render template color field 1`] = `
<DefaultConnectedControl
  component={[Function]}
  defaultValue="#112233"
  mapProps={
    Object {
      "defaultValue": [Function],
    }
  }
  model=".color"
/>
`;
